pkg/cloudprovider/providers/azure/azure_loadbalancer.go

```diff
@@ -157,14 +234,45 @@ func (az *Cloud) UpdateLoadBalancer(clusterName string, service *v1.Service, nod
 // have multiple underlying components, meaning a Get could say that the LB
 // doesn't exist even if some part of it is still laying around.
 func (az *Cloud) EnsureLoadBalancerDeleted(clusterName string, service *v1.Service) error {
-	lbName := getLoadBalancerName(clusterName)
-	pipName := getPublicIPName(clusterName, service)
+	isInternal := requiresInternalLoadBalancer(service)
+	lbName := getLoadBalancerName(clusterName, isInternal)
 	serviceName := getServiceName(service)
 
 	glog.V(2).Infof("delete(%s): START clusterName=%q lbName=%q", serviceName, clusterName, lbName)
 
-	// reconcile logic is capable of fully reconcile, so we can use this to delete
-	service.Spec.Ports = []v1.ServicePort{}
+	az.ensureLoadBalancerDeleted(clusterName, service, isInternal)
+
+	sg, existsSg, err := az.getSecurityGroup()
```

Comment:
We've got two Azure LoadBalancer Resource (internal, external), but only one Network Security Group (bound to vnet).

'ensureLoadBalancerDeleted' is for reconciliation of rules under certain Azure LoadBalancer resource.

If the service requests an external LoadBalancer:
Upon creation, controller calls it (isInternal=true) on internal load balancer resource to ensure the related internal rules deleted (in case the service was switched from internal load balancer type).
Then controller calls 'reconcileSecurityGroup' (wantLb=true) to ensure new rules under Network Security Group.

Upon deletion, controller calls it (isInternal=false) on external load balancer resource to ensure the related external rules deleted.
Then controller calls 'reconcileSecurityGroup' (wantLb=false) to cleanup related rules under Network Security Group.

For internal LoadBalancer, it'll be. creation: 'isInternal=false', 'wantLb=true'. Deletion: 'isInternal=true', 'wantLb=false'.

That is, the parameter value of 'isInternal' on 'ensureLoadBalancerDeleted', is orthogonal to 'wantLb' on 'reconcileSecurityGroup'. In this case, they may be not well suited to combine.

I think we may do some refactor later to wrap the similar code logic around 'reconcileSecurityGroup' calls, however.